src/libraries/System.Formats.Tar/tests/TarFile/TarFile.ExtractToDirectoryAsync.Stream.Tests.cs

```diff
@@ -329,5 +329,170 @@ public async Task UnseekableStreams_RoundTrip_Async(TestTarFormat testFormat)
             }
             while ((entry = await reader.GetNextEntryAsync()) != null);
         }
+
+        [Theory]
+        [MemberData(nameof(GetExactRootDirMatchCases))]
+        public async Task ExtractToDirectory_ExactRootDirMatch_RegularFile_And_Directory_Throws_Async(TarEntryFormat format, TarEntryType entryType, string fileName)
+        {
+            await ExtractToDirectory_ExactRootDirMatch_RegularFile_And_Directory_Throws_Internal_Async(format, entryType, fileName, inverted: false);
+            await ExtractToDirectory_ExactRootDirMatch_RegularFile_And_Directory_Throws_Internal_Async(format, entryType, fileName, inverted: true);
+        }
+
+        [Fact]
+        public async Task ExtractToDirectory_ExactRootDirMatch_Directory_Relative_Throws_Async()
+        {
+            string entryFolderName = "folder";
+            string destinationFolderName = "folderSibling";
+
+            using TempDirectory root = new TempDirectory();
+
+            string entryFolderPath = Path.Join(root.Path, entryFolderName);
+            string destinationFolderPath = Path.Join(root.Path, destinationFolderName);
+
+            Directory.CreateDirectory(entryFolderPath);
+            Directory.CreateDirectory(destinationFolderPath);
+
+            // Relative segments should not change the final destination folder
+            string dirPath1 = Path.Join(entryFolderPath, "..", "folder");
+            string dirPath2 = Path.Join(entryFolderPath, "..", "folder" + Path.DirectorySeparatorChar);
+
+            await ExtractRootDirMatch_Verify_Throws_Async(TarEntryFormat.Ustar, TarEntryType.Directory, destinationFolderPath, dirPath1, linkTargetPath: null);
+            await ExtractRootDirMatch_Verify_Throws_Async(TarEntryFormat.Ustar, TarEntryType.Directory, destinationFolderPath, dirPath2, linkTargetPath: null);
+        }
+
+        [ConditionalTheory(typeof(PlatformDetection), nameof(PlatformDetection.SupportsHardLinkCreation))]
+        [InlineData(TarEntryFormat.V7)]
+        [InlineData(TarEntryFormat.Ustar)]
+        [InlineData(TarEntryFormat.Pax)]
+        [InlineData(TarEntryFormat.Gnu)]
+        public async Task ExtractToDirectory_ExactRootDirMatch_HardLinks_Throws_Async(TarEntryFormat format)
+        {
+            await ExtractToDirectory_ExactRootDirMatch_Links_Throws_Async(format, TarEntryType.HardLink, inverted: false);
```

Comment:
It might be cleaner to have one method call with a foreach over new bool[] { true, false }